### PR TITLE
v2.x: Do not return MPI_ERR_PENDING from collectives.

### PR DESCRIPTION
**Pull request by abouteiller**
Collective operations cannot raise MPI_ERR_PENDING, only waitall, testall and friends are allowed to do so by spec. 

open-mpi/ompi@c98e97a
open-mpi/ompi@fbed6df

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1420/ for details.

**Comment by abouteiller** [comment]
open-mpi/ompi@c98e97a

**Comment by abouteiller** [comment]
bot:milestone:v2.x
bot:label:bug

**Comment by abouteiller** [comment]
added open-mpi/ompi@fbed6df
thanks @ggouaillardet

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1429/ for details.

**Comment by jsquyres** [comment]
@abouteiller @ggouaillardet Can one of you give us a final review thumbs up/down?

**Comment by hppritcha** [comment]
@ggouaillardet  please review

File: ompi/mca/coll/base/coll_base_barrier.c [hunk 1]
```diff
@@ -61,13 +61,14 @@ ompi_coll_base_sendrecv_zero(int dest, int stag,
...
     err = ompi_request_wait_all( 2, reqs, statuses );
-    if( MPI_ERR_IN_STATUS == err ) {
+    if( MPI_ERR_IN_STATUS == err ) { line = __LINE__;
         /* As we use wait_all we will get MPI_ERR_IN_STATUS which is not an error
          * code that we can propagate up the stack. Instead, look for the real
          * error code from the MPI_ERROR in the status.
          */
         int err_index = 0;
-        if( MPI_SUCCESS == statuses[0].MPI_ERROR ) {
+        if( MPI_SUCCESS == statuses[0].MPI_ERROR
```
Comment (ggouaillardet) [inline ompi/mca/coll/base/coll_base_barrier.c:70]: @abouteiller i do not fully understand the logic here.

with this code, i can see two cases that might not produce the desired output.

```
statuses[0].MPI_ERROR == MPI_ERR_PENDING
statuses[1].MPI_ERROR == MPI_SUCCESS
```

returns `MPI_SUCCESS`

```
statuses[0].MPI_ERROR == MPI_SUCCESS
statuses[1].MPI_ERROR == MPI_ERR_PENDING
```

returns `MPI_ERR_PENDING`

in these two cases, should the function returns an error code different than `MPI_SUCCESS` and `MPI_ERR_PENDING`

as a side note, is `MPI_Irecv` / `MPI_Isend` optimal here ?

for example, `MPI_Irecv` / `MPI_Send` could be a bit faster.

**Comment by ggouaillardet** [comment]
:bot:label:pushed-back

**Comment by jsquyres** [comment]
Due to lack of progress, we're pushing this to v2.0.1.

**Comment by hppritcha** [comment]
this can be addressed in 2.0.1

**Comment by abouteiller** [comment]
Sorry for some reason I missed the emails.

@ggouaillardet 
MPI_SUCCESS and MPI_ERR_PENDING from the same wait is an invalid outcome. If a request has MPI_ERR_PENDING in the status, at least one other request MUST have something other than MPI_SUCCESS or MPI_ERR_PENDING. Since there are only two here, that means the other has a non-trivial error status.

**Comment by jsquyres** [comment]
@ggouaillardet Are you happy with this now?  Do you want to remove the pushed-back label?

**Comment by ggouaillardet** [comment]
:+1:

**Comment by ggouaillardet** [comment]
:bot:nolabel:reviewed
Let me double check that again ...
iirc @bosilca made some optimizations in that area, and I want to make sure they do not conflict/override with this PR

**Comment by jsquyres** [comment]
bot:retest

**Comment by jsquyres** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1915/ for details.

**Comment by jsquyres** [comment]
@Di0gen @artpol84 @jladd-mlnx Is Mellanox Jenkins in a bad state?  It's now seeming to fail to clone any PR.

**Comment by hppritcha** [comment]
@ggouaillardet do you still see conflicts with @bosilca 's optimizations?

**Comment by ggouaillardet** [comment]
@hppritcha this is not really a conflict, but some changes were fixed/revamped/undone later.
i made #1273 with all the changes, and i'd rather bring them all at once.

@bosilca @abouteiller can you please advise ?

**Comment by jsquyres** [comment]
bot:mellanox:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1999/ for details.

**Comment by hppritcha** [comment]
moving to 2.1

**Comment by hppritcha** [comment]
@jsquyres should we close this PR and merge #1273 into 2.0.2 instead?

**Comment by jsquyres** [comment]
@hppritcha Sounds good.

**Pull request closed without merging**